### PR TITLE
fix(topology): restore edge detail and report every host behind a switch port

### PR DESCRIPTION
**Pull request by jpennin5**
## Description

Two commits, one subsystem — the second depends on the first's fetch
pattern, so they travel together.

**Commit 1 — edges had null ports and speeds:** the Integration API's
uplink object is only `{"deviceId": ...}` — it names the parent but
carries neither the ports a link lands on nor its negotiated speed. The
code read `portIndex` and `speedMbps` off that object, so those keys never
existed and every edge came back with null ports and null speed;
`type_detail` had the same problem. Fix: join the legacy `stat/device` and
`stat/sta` records on MAC, which do carry `port_idx`,
`uplink_remote_port`, `speed`, `full_duplex`, `sw_port` and
`wired_rate_mbps` (wireless rates are reported in kbps and converted).
Best-effort by design: a controller that refuses the legacy routes yields
a graph without port detail rather than no graph at all.

**Commit 2 — one host per port:** `get_port_mappings` keyed its map as
port → peer, so each write clobbered the last. One physical port routinely
carries many hosts — a virtualization host bridging its guests, or a
downstream unmanaged switch — and all but one vanished with no indication
anything had been dropped. Fix: each port maps to a **list** of peers, and
each peer carries `connected_name` so callers need not resolve node ids
themselves.

**Live verification (Network 10.5.67):** edges now carry real port
numbers and negotiated speeds; a switch port backing a virtualization
host reports every bridged guest instead of one.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

`get_port_mappings` port values change from a single object to a list —
previously the single object was an arbitrary survivor, so consumers were
already reading incomplete data.

## Testing

- Tests cover the legacy-record join (including kbps conversion and the
  degraded no-legacy-routes path) and multi-host ports; the multi-host
  test's mock dispatches on URL rather than call order.
- `pytest tests/unit/` and `pre-commit run --all-files` pass.
- `API.md` updated: `get_port_mappings` response example shows the
  list-of-peers shape with `connected_name`.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (API.md, CHANGELOG.md)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published (none — stands alone)

## AI Assistance

This PR was created with assistance from Claude Code.

**Human Review Status:** ✅ Reviewed and approved by @jpennin5
**Testing:** Full unit suite passing; edge detail and multi-host ports
verified against a live Network 10.5.67 controller.
**Security Review:** Completed — all fixture values are synthetic or from
documentation ranges.

### CODE REVIEW

File: src/tools/topology.py [hunk 1]
```diff
@@ -45,6 +45,52 @@ def _extract_device_detail(response: Any) -> dict[str, Any] | None:
...
+    Returns:
+        ``(devices_by_mac, clients_by_mac)``, both keyed by lowercased MAC
+    """
+
+    async def by_mac(endpoint: str) -> dict[str, dict[str, Any]]:
+        try:
+            response = await client.get(endpoint)
+        except (APIError, NetworkError) as exc:
+            client.logger.debug(f"No legacy detail from {endpoint}: {exc}")
+            return {}
```
[the remark this answers is not in the document]
Comment (jpennin5) [reply src/tools/topology.py:79]: Confirmed — `AuthenticationError` derives from UniFiMCPException, not APIError, so it did bypass the catch. Fixed in f8f674b: a 401/403 on the legacy routes now degrades like any other refusal, with a regression test.

File: src/tools/topology.py [hunk 2]
```diff
@@ -255,8 +309,10 @@ async def get_network_topology(
                     source_node_id=device_id,
                     target_node_id=uplink_device_id,
                     connection_type="uplink",
-                    source_port=uplink_info.get("portIndex"),
-                    speed_mbps=uplink_info.get("speedMbps"),
+                    source_port=legacy_uplink.get("port_idx"),
+                    target_port=legacy_uplink.get("uplink_remote_port"),
+                    speed_mbps=legacy_uplink.get("speed"),
+                    duplex="full" if legacy_uplink.get("full_duplex") else None,
                     is_uplink=True,
```
[the remark this answers is not in the document]
Comment (jpennin5) [reply src/tools/topology.py:324]: Fixed in f8f674b — `full_duplex=False` now reports `duplex="half"`; only an absent key yields null.

File: tests/unit/tools/test_topology_tools.py [hunk 3]
```diff
@@ -299,6 +299,44 @@ def mock_get_side_effect(url):
...
+        ]
+
+        with patch("src.tools.topology.UniFiClient") as mock_client:
+            mock_instance = mock_client.return_value.__aenter__.return_value
+            mock_instance.is_authenticated = True
+            mock_instance.resolve_site_id = AsyncMock(return_value="default")
+            mock_instance.settings = mock_settings
+            mock_instance.logger = MagicMock()
+            mock_instance.get = AsyncMock(side_effect=[[], clients, [], legacy_clients])
+
```
[the remark this answers is not in the document]
Comment (jpennin5) [reply tests/unit/tools/test_topology_tools.py:340]: Fixed in f8f674b — the mock now dispatches on URL, so the test no longer encodes the fetch order of a concurrent gather.

**Comment by enuno** [comment]
### Summary
Request changes. The PR is not yet merge-ready because the current status checks are failing.

### Critical findings
- Merge state: \ / \
- Failing check(s): Mermaid Diagram Sync Assistant
- The changed topology surface should be revalidated after the failing check is resolved.

### Suggestions
1. Resolve the failing status check(s) and rerun CI.
2. Reconfirm the changed topology behavior against the intended controller/fixture shapes.
3. Ensure regression tests cover the original failure mode and edge cases introduced here.
4. Update docs if any request/response or error semantics changed.

### Verification checklist for the author
- [ ] Required checks are passing
- [ ] Regression tests cover the changed path
- [ ] Lint / format / type checks pass locally
- [ ] Documentation is updated where user-facing behavior changed
- [ ] No secrets, credentials, or controller-identifying data were introduced

**Review by enuno — CHANGES_REQUESTED** [review]
Merge blocked until the failing status checks are resolved and the topology changes are revalidated.

**Comment by jpennin5** [comment]
The failing check is the `Mermaid Diagram Sync Assistant` GitHub App, not a workflow in this repository. Its check-run output on this PR reads: "Webhook Processing Failed — An unexpected error occurred while processing the webhook. Please try pushing a new commit or check your repository configuration." That processing happens on the app's side; pushing new commits to these PRs has re-triggered the check with the same result, so it is not reachable from a fork.

This is the loop documented in #112: the app has errored on every fork PR since late June, fork CI runs sit at `action_required` awaiting maintainer approval, and the daily automated review then reads the app error as a failing required check and requests changes.

Every check that can conclude on a fork PR is green here (both Socket Security checks). The remaining steps are repo-side — approving the CI runs and either repairing/removing the app integration or unrequiring its check. #117 merged with the identical check state, so the merge path exists. Happy to rebase or adjust anything code-side the moment CI can actually run.

**Comment by enuno** [comment]
Summary: request changes

Critical findings:
- Merge conflict / dirty merge state blocks safe review.

Suggestions:
1. Rebase onto main and resolve conflicts in: API.md, CHANGELOG.md, src/tools/topology.py, tests/unit/tools/test_topology_tools.py.
2. Re-run the touched test suite and pre-commit.
3. Re-check the feature behavior called out in the PR description.

Verification checklist:
- [ ] Rebasing completed cleanly
- [ ] Tests pass
- [ ] Lint/type checks pass
- [ ] Documentation remains aligned with implementation

**Review by enuno — CHANGES_REQUESTED** [review]
Automated review: merge conflicts/dirtiness detected (CONFLICTING/DIRTY). Please rebase and resolve before resubmitting.

**Comment by jpennin5** [comment]
Rebased onto current `main` (dd38e3f) — the conflict was in CHANGELOG, and it is resolved.

Resolution was gated rather than eyeballed: every CHANGELOG bullet already on `main` is verified still present (union, not overwrite), the sibling work merged into the shared files is verified intact, the full unit suite passes, and pre-commit passes for every file the PR touches.

Codecov's remaining line is covered: the non-dict row guard in the legacy stat join.

GitHub now reports this PR mergeable.

**Comment by enuno** [comment]
## Summary
Request changes.

## Critical findings
- CI is not green on this PR.
- Failing checks: Mermaid Diagram Sync Assistant; Pre-commit Hooks; Secret Scanning; Security Scan Summary.
- Touched scope: API.md, CHANGELOG.md, src/tools, tests/unit.

## Suggestions
1. Fix the failing checks on the current branch and rerun them until the gate is green.
2. Re-verify the changed code paths and tests for the touched scope above.
3. If any failing security or lint checks are expected, document the rationale and follow-up needed.

## Verification checklist
- [ ] `pre-commit run --all-files` passes
- [ ] Security/lint checks are green or explicitly justified
- [ ] Relevant unit tests cover the changed behavior
- [ ] No unintended API or documentation regressions

**Review by enuno — CHANGES_REQUESTED** [review]
CI is red on this PR (Mermaid Diagram Sync Assistant; Pre-commit Hooks; Secret Scanning; Security Scan Summary). Please resolve the failing checks before merge.
